### ISSUE TITLE
create human-readable script for integration test and demo

### ISSUE THREAD

**Issue by ahdinosaur** [root-systems/cobuy#207]
this ticket is to create a human-readable script for how to start with a fresh load of Cobuy and move through the various stages of the app.

this should include reasonable demo data, that matches what we'd expect in a real-world use case.

this serves two purposes:

- it provides clear instructions for how to test the app, manually as a human
- it provides clear instructions for how to demo the app, as a presentation

we can later turn these manual human instructions into automatic integration tests. (TODO make ticket for this)

**Comment by iainkirkpatrick** [comment]
first bash at a high-level outline for a script (using the current names for processes, some probably should be changed):

- home screen
- register
- user name and password
- create my profile
- start an order
- finish prereqs
- create group profile
- invite 2 more members
- create supplier profile
- create 3 products, each with 2 price specs
- start order
- select the group
- select the supplier
- select myself as admin

(new browser)
- sign in as another member
- create profile
- cast intents in open order at same time as original member is casting intents (i.e. demoing realtime)
- order closes (how? timer that we skip to in video?)
- admin sees order complete invoice
- admin sees order complete sheet breakdown by group member

**Comment by iainkirkpatrick** [comment]
is there any issue with using our real names / email addresses in the demo? i guess it'd be cool to show the real email for an invited member, but wouldn't need to show an actual inbox etc could just cut the video so it shows the email content.

**Comment by iainkirkpatrick** [comment]
@agentlewis what would be a good name for a demo Group? i.e. is it a community of learning that would be the level of a group in Tapin? i.e. 'Test COL' or something more exciting

**Comment by iainkirkpatrick** [comment]
some draft data for a supplier, changes welcome:

```
{
  name: "School Supplies R'Us",
  description: "If you're yearning for some learning, we'll set your mind churning"
  products: [
    {
      name: "Laptop Pro 3",
      description: "The latest and greatest for passing the tough test",
      imageLink: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcR71MqHHl8WFQ-Llf3EEJJSnBMKDRKqxyoWr0DMXEj0RgSrpS0W"
      priceSpecs: [
        {
          minimum: 10,
          price: 500,
          currency: 'NZD'
        },
        {
          minimum: 100,
          price: 300,
          currency: 'NZD'
        }
      ]
    },
    {
      name: "Ergo-Chair 9000"
      description: "Relax your back - why not buy a stack?"
      imageLink: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcTcZGA_ssdCo0LgwYGkR-QT03hegiqELsyPgB2drk6PzS6zV8uX"
      priceSpecs: [
        {
          minimum: 100,
          price: 100,
          currency: 'NZD'
        },
        {
          minimum: 500,
          price: 50,
          currency: 'NZD'
        }
      ]
    },
    {
      name: "The AllDesk"
      description: "Just the right height for students and teachers alike. Also comes in dark or light"
      imageLink: "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS0GW5TvkcWjaLTDg2-vQPcL5j5r9igWASYzSerQh_eQ2cVuyqY"
      priceSpecs: [
        {
          minimum: 100,
          price: 250,
          currency: 'NZD'
        },
        {
          minimum: 1000,
          price: 120,
          currency: 'NZD'
        }
      ]
    }
  ]
}
```

**Comment by danalexilewis** [comment]
@iainkirkpatrick Ruru Kāhui Ako

**Comment by iainkirkpatrick** [comment]
@agentlewis is that a real kahui ako? if so, should we have a real description, phone, email, website for them?

**Comment by iainkirkpatrick** [comment]
![image uploaded from ios 1](https://user-images.githubusercontent.com/1994975/31867585-f2d99306-b7ee-11e7-9630-573e41a4411c.jpg)

**Comment by iainkirkpatrick** [comment]
Had a chat with @agentlewis around video content, which spiralled a little into what things down the road for Cobuy / Cosupply and their Tapin cousins might look like also.

Dan's explanation of the video is that it needs to cover both:
- progress we have made so far with the money
- a pitch for the direction of Tapin overall, i.e. selling the vision

As per the image above, Dan suggested that the Unique Selling Points of Cobuy are:

- real-time intent casting
- order summaries
- task management

So for the video to hit these USP's, while showing progress and contributing towards advertising the dream, we suggested that we should show:

- an admin configuring an order (i.e. selecting the group, the supplier, and the admin (facilitator?) of the order
- that admin starting the order, which assigns the task to cast intents to 3 other members of the group. These users would all have a browser window each open (i.e. 4 browsers open including the admin's, each in a quadrant of the screen), and the video would show the task appearing on each dashboard.
- each user casting some intents, which then are shown on every other screen updating in real-time
- [two weeks later] the order closes, and an order summary is shown to the admin / facilitator of the order.

We talked about what happens after that - Dan suggests that people will be interested in the 'whole story' of group purchasing, which includes confirmation from the supplier, the group members paying etc. We think that in the end, there is probably space for a 'Tapin Supply' - a version of the idea of CoSupply that has been floated around at various points over the last year or so. This would be software that the suppliers use to manage Cobuy orders at it's lightest - an account management application - the heavy version would have that as well as inventory management etc. The alternative to this is leaning further on the task management system in Cobuy to assign tasks to admins / users to do outside of Cobuy, basically like reminders - i.e. 'confirm the supplier has received your order through emailing you', 'confirm everyone has paid'.

@agentlewis anything else I've missed?

**Comment by danalexilewis** [comment]
@iainkirkpatrick - no that’s really good.

If we can get the video down to 30-60 seconds that would be good.

May mean accelerorating the playback a little (certainly any typing)

**Comment by iainkirkpatrick** [comment]
20-30 secs max, maybe a gif

**Comment by iainkirkpatrick** [comment]
FYI all i'm jamming on some seed data for the tapin demo so that we don't need to have email invites working to have other accounts that belong to a single group (after talking with @Julietbc and @sarah-arrrgh )

**Comment by iainkirkpatrick** [comment]
#251 / #252 as PR / ticket for seed data

**Comment by sarah-arrrgh** [comment]
- seed data will exist in db once the seeds have been run

- 20-30 secs total

---

- log in as dan to start an order
- others are already members of group
- we are not showing setting up a group / inviting members
- so we're just showing dan starting an order from order config page
- this is the group this order is for
- this is the supplier it's for
- this is the admin (can choose any member of group, will choose himself)

REAL TIME 4 BROWSER WINDOWS OPEN

- click start order
- order is set up, hooray!

- assign all members of group a task to cast their intents (check this is working)
- show each member casting their intent in their browser window (juliet currently getting this working #39)
- show that when someone casts intent it updates group total order (not currently doing this)

2 WEEKS LATER

BACK TO ONE SCREEN

- order closes
- see order summary as an admin (not currently working, may need to fake it in storybook)

**Comment by sarah-arrrgh** [comment]
bug: start order creates a blank group and blank supplier in pre-reqs

**Comment by sarah-arrrgh** [comment]
Current app (master branch) status things to note:

- db seeding works good - run `burnthedb` when pulling down master to seed the db (and to reset the seeding)
- long delay between hitting *start order* and the task *complete your order* coming up (speed this up for video)
- start recording straight from *start order* view
- on click *done*, video magic back to dash (as current bug = prereq tasks need to be completed despite there already being group + members + supplier setup)
- clicking *cast your intent for order*  loads a view with no tasks on it - assume what @Julietbc is currently working on will fix this?

**Comment by iainkirkpatrick** [comment]
this is done now, but wondering if we should extract the steps into another doc somewhere to be manual tests / integration tests... thoughts @ahdinosaur ?

**Comment by ahdinosaur** [comment]
@iainkirkpatrick yeah, before this ticket is "done" i reckon we should have a TEST.md or wiki page with the human-readable scripts to manually test the app.

**Comment by ahdinosaur** [comment]
hey @iainkirkpatrick @sarah-arrrgh, what's the status on this ticket? seems y'all did a heap of work on this for the demo, but wonder where this leaves the original deliverable described in the ticket for a human-readable script we can use to manually test the app from start to finish, maybe including some good edge cases. already it's been very valuable having this for the demo in surfacing bugs, would love for us to be (manually) testing the app before every release, and to start releasing at the end of every sprint.

**Comment by sarah-arrrgh** [comment]
TODO: put in the wiki, how we would manually run a test, so when we have a new release, we can run through the script and make sure everything works.

how do i start from scratch
burnthedb
.
.
create an account etc
logout
login
.
.

**Comment by sarah-arrrgh** [comment]
Link to wiki page:
https://github.com/root-systems/cobuy/wiki/Human-readable-script

End of day report:
- have started script (link above)
- blocked by not actually being able to select the group and supplier etc that i created to finish off the order setup task

**Comment by sarah-arrrgh** [comment]
Bugs:

- [ ] adding members to a group you're setting up - some aren't saved... very glitchy. profiles aren't saved to db / names disappear once invited,
- [ ] on start order during pre-reqs phase - one or all of the groups/suppliers/admins i just set up wont appear in the drop down menu (usually blank)

**Comment by sarah-arrrgh** [comment]
clearing browser cache + burning them ALL + burnthedb fixes above bugs

**Comment by sarah-arrrgh** [comment]
Bugzzz: 

- [ ] Can't log in/sign up as a member already added to a group and place an order on an order that's already been set up.

- [ ] And can't add a member to a new group if they already have an account from the db seed file

Seems like after burningthemall and clearing cache, the app will work long enough to set up a new group, supplier and order, but then when i log out and try to log in as someone else, or even myself again, it's all lost? db records are deleted somewhere along the way?

**Comment by danalexilewis** [comment]
@sarah-arrrgh I think we should break these bugs out into seperate tickets - otherwise it appears like you are working on all of them.

I would also suggest writing each bug as a series of steps to replicate it. Gherkin works really good for this:

 - Given I am a user who has been added to an order
 - And I have logged in
 - And I am looking at the order task
 - And I have selected a product 
 - And I have entered my intent
 - When I click save
 - Then I am taken back to the Oder task view
 - **But nothing happens**

This makes a very clear set up instructions that any dev can pickup and replicate the bug off of. Not this is similar to the codecept library :)
